cmd/kubeadm/app/phases/addons/dns/dns.go

```diff
@@ -297,6 +298,83 @@ func createDNSService(dnsService *v1.Service, serviceBytes []byte, client client
 	return nil
 }
 
+func createCoreDNSConfigMap(client clientset.Interface, cm *v1.ConfigMap) error {
+	// Get the CoreDNS ConfigMap. If not present, Create a new default ConfigMap.
+	if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Get(kubeadmconstants.CoreDNSConfigMap, metav1.GetOptions{}); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil
+		}
+		if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Create(cm); err != nil {
+			if !apierrors.IsAlreadyExists(err) {
+				return errors.Wrap(err, "unable to create configmap")
+			}
+		}
+	}
+	// Get the Corefile and installed CoreDNS version.
+	corefile, currentInstalledCoreDNSVersion, err := GetCoreDNSInfo(client)
+	if err != nil {
+		return errors.Wrap(err, "unable to fetch CoreDNS current installed version and ConfigMap.")
+	}
+	// Check if the current Configuration present is the default (unmodified) version.
+	// If not, try and migrate the Config.
+	if !migration.Default("", corefile) {
+		updatedCorefile, err := migration.Migrate(currentInstalledCoreDNSVersion, kubeadmconstants.CoreDNSVersion, corefile, false)
+		if err != nil {
+			return errors.Wrap(err, "unable to migrate CoreDNS Configmap")
+		}
+
+		if _, err := client.CoreV1().ConfigMaps(cm.ObjectMeta.Namespace).Update(&v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      kubeadmconstants.CoreDNSConfigMap,
+				Namespace: metav1.NamespaceSystem,
+			},
+			Data: map[string]string{
+				"Corefile":          updatedCorefile,
+				"Corefile-previous": corefile,
```

Comment:
the Corefile itself is not that much data, but i wonder if this is the best backup option. the alternative is to write to file, which seems slightly better to me, but the question is where.

i'd like to hear more opinions on this.

Reply:
> the Corefile itself is not that much data, but i wonder if this is the best backup option. the alternative is to write to file, which seems slightly better to me, but the question is where.
> i'd like to hear more opinions on this.

FWIW, my biased opinion ... Keeping the old version "in kubernetes" (e.g. in the configmap) seems more intuitive to me, from a user perspective. The old Corefile is readily available to roll back to, or to see as reference in the event that the migrated file is not working as expected.  In the ConfigMap, it's already in a place that any user that has permission to edit the configmap can access (same configmap).  If saved to a file (i assume to the local filesystem), it would only be saved to where ever kubeadm was executed.

Reply:
Yes, I agree completely with @chrisohaver

Reply:
> The old Corefile is readily available to roll back to

does kubeadm do anything to rollback coredns versions? if so, does it also handle putting the "previous" data back as the current config? the "previous" key seems underspecified to be usable in an automated process... nothing indicates what version it is intended to work with. If I'm reading this correctly, even patch release upgrades would move current config into the previous key, right?